x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java

```diff
@@ -225,25 +225,43 @@ public static <T> T extractOptionalEnum(
         Map<String, Object> map,
         String settingName,
         String scope,
-        CheckedFunction<String, T, IllegalArgumentException> converter,
-        T[] validTypes,
+        EnumConstructor<T> constructor,
+        T[] validValues,
         ValidationException validationException
     ) {
         var enumString = extractOptionalString(map, settingName, scope, validationException);
         if (enumString == null) {
             return null;
         }
 
-        var validTypesAsStrings = Arrays.stream(validTypes).map(type -> type.toString().toLowerCase(Locale.ROOT)).toArray(String[]::new);
+        var validValuesAsStrings = Arrays.stream(validValues).map(type -> type.toString().toLowerCase(Locale.ROOT)).toArray(String[]::new);
         try {
-            return converter.apply(enumString);
+            var createdEnum = constructor.apply(enumString);
+            validateEnumValue(createdEnum, validValues);
+
+            return createdEnum;
         } catch (IllegalArgumentException e) {
-            validationException.addValidationError(invalidValue(settingName, scope, enumString, validTypesAsStrings));
+            validationException.addValidationError(invalidValue(settingName, scope, enumString, validValuesAsStrings));
         }
 
         return null;
     }
 
+    private static <T> void validateEnumValue(T enumValue, T[] validValues) {
+        if (Arrays.asList(validValues).contains(enumValue) == false) {
```

Comment:
Can you achieve the same by passing an `EnumSet<T>` instead of `T[] validValues`? It would make the lookup easier.

Nit: in the docs a generic enum parameter is referred to as `E`

https://docs.oracle.com/javase/8/docs/api/java/util/EnumSet.html